server/leafnode.go

```diff
@@ -0,0 +1,1074 @@
+	c.mu.Unlock()
+	s.mu.Lock()
+	delete(s.leafs, cid)
+	s.mu.Unlock()
+}
+
+type leafConnectInfo struct {
+	JWT  string `json:"jwt,omitempty"`
+	Sig  string `json:"sig,omitempty"`
+	User string `json:"user,omitempty"`
+	Pass string `json:"pass,omitempty"`
+	TLS  bool   `json:"tls_required"`
+	Comp bool   `json:"compression,omitempty"`
+	Name string `json:"name,omitempty"`
+
+	// Just used to detect wrong connection attempts.
+	Gateway string `json:"gateway,omitempty"`
+}
+
+// processLeafNodeConnect will process the inbound connect args.
+// Once we are here we are bound to an account, so can send any interest that
+// we would have to the other side.
+func (c *client) processLeafNodeConnect(s *Server, arg []byte, lang string) error {
+	// Way to detect clients that incorrectly connect to the route listen
+	// port. Client provided "lang" in the CONNECT protocol while LEAFNODEs don't.
+
+	if lang != "" {
+		c.sendErrAndErr(ErrClientConnectedToLeafNodePort.Error())
+		c.closeConnection(WrongPort)
+		return ErrClientConnectedToLeafNodePort
+	}
+
+	// Unmarshal as a leaf node connect protocol
+	proto := &leafConnectInfo{}
+	if err := json.Unmarshal(arg, proto); err != nil {
+		return err
+	}
+
+	// Reject if this has Gateway which means that it would be from a gateway
+	// connection that incorrectly connects to the Route port.
+	if proto.Gateway != "" {
+		errTxt := fmt.Sprintf("Rejecting connection from gateway %q on the Route port", proto.Gateway)
+		c.Errorf(errTxt)
+		c.sendErr(errTxt)
+		c.closeConnection(WrongGateway)
+		return ErrWrongGateway
+	}
+
+	// Leaf Nodes do not do echo or verbose or pedantic.
+	c.opts.Verbose = false
+	c.opts.Echo = false
+	c.opts.Pedantic = false
+
+	// Create and initialize the smap since we know our bound account now.
+	c.initLeafNodeSmap()
+
+	// We are good to go, send over all the bound account subscriptions.
+	s.startGoRoutine(func() {
+		c.sendAllAccountSubs()
+		s.grWG.Done()
+	})
+
+	// Add in the leafnode here since we passed through auth at this point.
+	s.addLeafNodeConnection(c)
+
+	// Announce the account connect event for a leaf node.
+	// This will no-op as needed.
+	s.sendLeafNodeConnect(c.acc)
+
+	return nil
+}
+
+// Snapshot the current subscriptions from the sublist into our smap which
+// we will keep updated from now on.
+func (c *client) initLeafNodeSmap() {
+	acc := c.acc
+	if acc == nil {
+		c.Debugf("Leaf node does not have an account bound")
+		return
+	}
+	// Collect all subs here.
+	_subs := [256]*subscription{}
+	subs := _subs[:0]
+	ims := []string{}
+	acc.mu.RLock()
+	acc.sl.All(&subs)
+	// Since leaf nodes only send on interest, if the bound
+	// account has import services we need to send those over.
+	for isubj := range acc.imports.services {
+		ims = append(ims, isubj)
+	}
+	acc.mu.RUnlock()
+
+	// Now walk the results and add them to our smap
+	c.mu.Lock()
+	for _, sub := range subs {
+		// We ignore ourselves here.
+		if c != sub.client {
+			c.leaf.smap[keyFromSub(sub)] += 1
```

Comment:
Didn't know we could do that.. which is great. Replace with `c.leaf.smap[keyFromSub(sub)]++` instead to avoid a lint report.

Reply:
Thx will fix